src/handleInteraction.ts

```diff
@@ -0,0 +1,193 @@
+			if (viaDm) {
+				// We need to say *something* to the interaction itself, or Discord will think we died.
+				const content = ':paperclip: Check your DMs';
+				if (interaction.deferred) {
+					try {
+						await interaction.editReply(content);
+					} catch (error) {
+						logger.error('Failed to edit reply to interaction:', error);
+					}
+				} else {
+					try {
+						await interaction.reply({ content, ephemeral: true });
+					} catch (error) {
+						logger.error('Failed to reply to interaction:', error);
+					}
+				}
+			}
+			if (interaction.deferred && !viaDm) {
+				try {
+					if (typeof options === 'string') {
+						await interaction.followUp({ ephemeral: true, content: options });
+					} else {
+						await interaction.followUp({ ...options, ephemeral: true });
+					}
+				} catch (error) {
+					logger.error('Failed to follow up on interaction:', error);
+				}
+			} else {
+				const reply = await replyPrivately(interaction, options, viaDm);
+				if (reply === false) {
+					logger.info(`User ${logUser(interaction.user)} has DMs turned off.`);
+				}
+			}
+		},
+		async reply(options) {
+			if (interaction.deferred) {
+				try {
+					await interaction.editReply(options);
+				} catch (error) {
+					logger.error('Failed to edit reply to interaction:', error);
+					await interaction.followUp(options);
+				}
+			} else {
+				try {
+					if (typeof options === 'string') {
+						await interaction.reply(options);
+					} else if (
+						!('shouldMention' in options) ||
+						options.shouldMention === undefined ||
+						options.shouldMention
+					) {
+						// Doesn't say whether to mention, default to `true`
+						await interaction.reply(options);
+					} else {
+						// Really shouldn't mention
+						await interaction.reply({
+							...options,
+							allowedMentions: { users: [] },
+						});
+					}
+				} catch (error) {
+					logger.error('Failed to reply to interaction:', error);
+				}
+			}
+
+			if (typeof options !== 'string' && 'ephemeral' in options && options?.ephemeral === true) {
+				// FIXME: We didn't actually send the reply if we errored out
+				logger.info(
+					`Sent ephemeral reply to User ${logUser(interaction.user)}: ${JSON.stringify(options)}`
+				);
+			}
+		},
+		async followUp(options) {
+			if (
+				typeof options !== 'string' &&
+				(!('reply' in options) || options.reply === false || options.reply === undefined) &&
+				interaction.channel?.isTextBased() === true
+			) {
+				return (
+					(await sendMessageInChannel(interaction.channel, { ...options, reply: undefined })) ??
+					false
+				);
+			}
+			try {
+				return await interaction.followUp(options);
+			} catch (error) {
+				logger.error('Failed to follow up on interaction:', error);
+				return false;
+			}
+		},
+		sendTyping() {
+			void interaction.channel?.sendTyping();
+			logger.debug(
+				`Typing in channel ${interaction.channel?.id ?? 'nowhere'} due to Context.sendTyping`
+			);
+		},
+	};
```

Comment:
This context object declaration is really lengthy. It sorta distracts from the purpose of `handleInteraction` because most of the logic is not even used here. Plus, it gets generated every single time an interaction is received.

Is there any way to move this logic out of this file and have it be loaded in as an object (and edited) at runtime? It would really clean up this file. And then we can edit the command context whenever necessary without messing with this event handler.

Reply:
Good idea. I think each of the methods could be abstracted into separate files, sortof like the `replyPrivately` and `sendMessageInChannel` functions are. Those two helpers could probably be combined with the context methods somewhat.

I think constructing the context is part of the purview of "handling the interaction," as part of that job is to select the appropriate command and execute it, which necessarily involves a context specific to the invocation. The methods tho could definitely be defined with factory functions elsewhere, since they only need the current interaction and the logger instance to know what to do.

That work could be done in a separate PR tho. Once this PR is in, then the team can work on various parts of the bot, including commands and even the context construction. This PR mainly lays down the initial bones and interface joints for the whole thing.

Reply:
IMO length itself is not a huge concern, but when the long function starts dealing with things that it should not need to deal with, like the ins and outs of Discord's Interactions API, that's not ideal.

I think factory functions might be the way to go here, tho regardless of how we solve this problem, I think the solution can be non-breaking enough to warrant a separate PR.

Reply:
I've created #21 to dive into this further, if this needs more discussion.